Write the entirety of the Wiimote EEPROM, in a per-Wiimote file

Pokechu22 opened the pull request:
Previously, only Mii data was written.  Additionally, the file containing mii data was shared for all Wiimotes, which made it a lot less useful.

Additionally, the file was read/written on each Wiimote read, even though the whole EEPROM was kept in memory.  This was bad for performance and not particularly necessary (it did enforce that the data was properly shared between all Wiimotes, but that's not something I want).

-----

Some notes about this:

* If there is no `Wiimote#.bin` file, the previous `mii.bin` file will be read (if it exists) and used to populate the mii block.  However, since the `Wiimote#.bin` files are not written until a write to EEPROM happens, this will continue happening on each startup.  Also, the `mii.bin` file isn't ever deleted, which is probably fine but might get weird if all wiimotes have bin files with them since the mii file isn't used at that point.
* I think it would be neat to have files not tied directly to Wiimote numbers, since you might want to save more data per game (or data from different games or whatever).  But I've got no idea how to do that from a UI perspective.
* It would be neat to import (and _maybe_ export, though that may be dangerous) data to physical Wiimotes, but that does depend on having further management of the files than just tying it to the Wiimote number.
* This includes writing EEPROM for the balance board, which is ~~[currently unknown](https://wiibrew.org/wiki/Wii_Balance_Board#EEPROM_Memory).  Technically Dolphin allowed reads and writes to it before, but it would only interact with the same `mii.bin` file.  Now, a write to the EEPROM for the wii balance board will create a `Wiimote5.bin` file (I think), which might be incorrect.~~  correct based on hardware testing, and will go into a `BalanceBoard.bin` file.  That said, this only happens after a write, which presumably no actual game does.
* Data might be written too infrequently now.  It's written when `Reset()` is called while the eeprom is dirty, which should include disconnects; however, that only happens when the game is closed eventually.  If there's a crash or something in the middle, the file remains unwritten.

mbc07 commented:
About exporting emulated Wiimote EEPROM to real Wiimote EEPROM: that's a bad idea as the user accessible area also store calibration data for the motion sensors and that's unique on each unit, overwriting these values will cause issues. If this get implemented, make sure to always skip the calibration data when exporting from emulated => real Wiimotes...

jordan-woyak commented on Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp:
```diff
@@ -74,34 +75,69 @@ void Wiimote::Reset()
+    // Write out existing EEPROM
+    std::ofstream file;
+    File::OpenFStream(file, eeprom_file, std::ios::binary | std::ios::out);
+    file.write((char*)m_eeprom.data.data(), EEPROM_FREE_SIZE);
```
This should be a `reinterpret_cast`.

jordan-woyak commented on Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp:
```diff
@@ -74,34 +75,69 @@ void Wiimote::Reset()
+    // Read existing EEPROM
+    std::ifstream file;
+    File::OpenFStream(file, eeprom_file, std::ios::binary | std::ios::in);
+    file.read((char*)m_eeprom.data.data(), EEPROM_FREE_SIZE);
```
`reinterpret_cast`

jordan-woyak commented on Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp:
```diff
@@ -136,7 +172,7 @@ void Wiimote::Reset()
 }
 
-Wiimote::Wiimote(const unsigned int index) : m_index(index)
+Wiimote::Wiimote(const unsigned int index) : m_index(index), m_eeprom_dirty(false)
```
You could just initialize this value in the class definition: `bool m_eeprom_dirty = false;`.

jordan-woyak commented on Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp:
```diff
@@ -74,34 +75,69 @@ void Wiimote::Reset()
+      // Import from the existing mii.bin file, if present
+      std::ifstream file;
+      File::OpenFStream(file, mii_file, std::ios::binary | std::ios::in);
+      file.read((char*)m_eeprom.mii_data_1.data(), m_eeprom.mii_data_1.size());
```
`reinterpret_cast`

Pokechu22 commented:
Another thing I just remembered: Save-states already did store the entirety of the EEPROM.  However, loading a state generally wouldn't cause issues, as the mii data was re-read from disk each time so the loaded one was ignored.  (The only case that could possibly happen is if it wrote without reading, which would only happen when saving on one of the transfer screens (e.g. sending miis to the remote).  Now, since it loads on startup and saves on exit, a savestate will overwrite the previously stored data.  I don't know if that's a problem (and I think that might already be the case for normal save files).


And regarding writing to a physical remote: Yeah, I'm aware of that.  It would make sense to allow configuring which sections to write, with the Mii blocks and the user data block enabled by default and the others disabled with a warning on enabling them.  Being able to chose to only write the user block seems useful, and making the rest also configurable would be nice.

mbc07 commented:
> And regarding writing to a physical remote: Yeah, I'm aware of that. It would make sense to allow configuring which sections to write, with the Mii blocks and the user data block enabled by default and the others disabled with a warning on enabling them. Being able to chose to only write the user block seems useful, and making the rest also configurable would be nice.

Honestly, (over)writing the calibration data should **never** be allowed if this get implemented. I already see users ignoring any warning and then complaining their Wiimotes are broken on the forums...

Pokechu22 commented:
> Honestly, (over)writing the calibration data should **never** be allowed if this get implemented. I already see users ignoring any warning and then complaining their Wiimotes are broken on the forums...

Fair enough.  However, that's something that will be handled in a later PR; this one doesn't do anything with physical remotes.  For now I think it's best to just not worry about it; when someone does something with physical remotes then we can worry about it.  (I lack the hardware to actually do any testing with them in Dolphin).

stenzek commented:
Seems this needs a rebase now. If you can do that I'll go ahead and merge.

Helios747 commented:
@dolphin-emu-bot rebuild

JosJuice commented:
Will the EEPROM get saved to disk when doing a forced shutdown of the emulated Wii? (I can see that Reset gets called when disconnecting Wii Remotes, which happens when doing a clean shutdown, but I can't find any code that calls Reset on all shutdowns.)

Just to clarify, I'm not expecting any support for the case where Dolphin crashes, but it would be nice if we could handle the case where the emulated software crashes (and the case of using Android, since it currently doesn't support clean shutdowns...)

Pokechu22 commented:
Looks like it doesn't get saved in that case.  It also doesn't get saved in the case of launching a Gamecube game from the Wii Menu and then stopping it (which seems odd, as normally wiimotes are disconnected in that case).  On the other hand, it doesn't get *loaded* either, so if the Wii Menu is launched again and then closed, it should be saved (which is a bit jank).

Pokechu22 commented:
I've added code to reset Wiimotes in both of those cases.  That said, I'm not sure it's the best way of doing it.  I don't have the ability to use real Wiimotes with Dolphin, so I'm not sure if this will properly disconnect them (or if they *should* be disconnected when using Dolphin).

JosJuice commented:
The call you added to `ResetAllWiimotes` in DolphinQt won't run on Android or no-gui. Could you put it in Core code instead?

Pokechu22 commented:
Done.  I've also added a special case to `Wiimote::GetName` for the balance board (though it won't come up in practice due to no games being known to use it; I did test and confirm that the balance board does have usable EEPROM though).  I also added a log message when the EEPROM is written (only when it is dirty).

JosJuice commented:
@dolphin-emu-bot rebuild
